File: backend/models/trace.py
# backend/models/trace.py
from pydantic import BaseModel, Field
from typing import Optional, Any
from datetime import datetime
import uuid


class AgentTrace(BaseModel):
    """
    Records every agent decision for the Trace Viewer in the Flutter app.
    This is what makes CIRO's reasoning visible — the key differentiator.
    """
    trace_id: str = Field(default_factory=lambda: f"tr_{uuid.uuid4().hex[:8]}")
    incident_id: str
    agent: str  # "classification_agent", "resource_allocation_agent", etc.
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    
    input_summary: str   # Human-readable description of what the agent received
    reasoning: str       # The agent's reasoning (from LLM or rule-based logic)
    output: dict         # The structured output produced
    
    decision: str        # e.g., "CLASSIFY_AND_RESPOND", "FLAG_FOR_VERIFICATION"
    next_agent: Optional[str] = None  # Which agent runs next
    
    latency_ms: int = 0
    model: Optional[str] = None  # "gemini-1.5-flash", "rule-based", etc.
    is_degraded: bool = False     # True if LLM failed and rules were used
    
    class Config:
        json_encoders = {datetime: lambda v: v.isoformat()}

File: ciro-mobile/lib/data/models/trace.dart
class AgentTrace {
  final String traceId;
  final String incidentId;
  final String agent;
  final DateTime timestamp;
  final String inputSummary;
  final String reasoning;
  final Map<String, dynamic> output;
  final String decision;
  final String? nextAgent;
  final int latencyMs;
  final String? model;
  final bool isDegraded;

  AgentTrace({
    required this.traceId,
    required this.incidentId,
    required this.agent,
    required this.timestamp,
    required this.inputSummary,
    required this.reasoning,
    required this.output,
    required this.decision,
    this.nextAgent,
    this.latencyMs = 0,
    this.model,
    this.isDegraded = false,
  });

  factory AgentTrace.fromJson(Map<String, dynamic> json) {
    return AgentTrace(
      traceId: json['trace_id'] as String,
      incidentId: json['incident_id'] as String,
      agent: json['agent'] as String,
      timestamp: DateTime.parse(json['timestamp'] as String),
      inputSummary: json['input_summary'] as String,
      reasoning: json['reasoning'] as String,
      output: Map<String, dynamic>.from(json['output'] ?? {}),
      decision: json['decision'] as String,
      nextAgent: json['next_agent'] as String?,
      latencyMs: json['latency_ms'] as int? ?? 0,
      model: json['model'] as String?,
      isDegraded: json['is_degraded'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() => {
    'trace_id': traceId,
    'incident_id': incidentId,
    'agent': agent,
    'timestamp': timestamp.toIso8601String(),
    'input_summary': inputSummary,
    'reasoning': reasoning,
    'output': output,
    'decision': decision,
    'next_agent': nextAgent,
    'latency_ms': latencyMs,
    'model': model,
    'is_degraded': isDegraded,
  };
}
